### PR TITLE
expression: fix compatible problem with mysql when parse datetime

### LINKED ISSUE #18309: cast string to datetime is not compatible with mysql 5.7
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
tidb(localhost:4000) > select cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime);
+--------------------------------------------------------------+
| cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime) |
+--------------------------------------------------------------+
| NULL                                                         |
+--------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

### 2. What did you expect to see? (Required)
in mysql 5.7:
```sql
mysql > select cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime);                                                                                                                                                                                         
+--------------------------------------------------------------+
| cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime) |
+--------------------------------------------------------------+
| 2020-06-01 00:00:00                                          |
+--------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql > show warnings;
+---------+------+--------------------------------------------------------------------+
| Level   | Code | Message                                                            |
+---------+------+--------------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect datetime value: '2020-06-01 00:00:00 00:00:00' |
+---------+------+--------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
in TiDB:
```sql
tidb(localhost:4000) > select cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime);
+--------------------------------------------------------------+
| cast(concat("2020-06-01 00:00:00", " 00:00:00") as datetime) |
+--------------------------------------------------------------+
| NULL                                                         |
+--------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

tidb(localhost:4000) > show warnings;
+---------+------+----------------------------------------------------------+
| Level   | Code | Message                                                  |
+---------+------+----------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '2020-06-01 00:00:00 00:00:00' |
+---------+------+----------------------------------------------------------+
1 row in set (0.01 sec)
```

### 4. Affected version (Required)

v4.0.1

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #18309

Problem Summary: fix compatible problem with mysql when parse datetime

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- fix compatible problem with mysql when parse datetime

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -786,7 +786,15 @@ func isValidSeparator(c byte, prevParts int) bool {
 		return true
 	}
 
-	return prevParts == 2 && (c == ' ' || c == 'T')
+	if prevParts == 2 && (c == ' ' || c == 'T') {
+		return true
+	}
+
+	if prevParts > 4 && !isDigit(c) {
```
Comment (SunRunAway) [inline types/time.go:793]: Please document and give an example for this situation.

**Review by wshwsh12 — COMMENTED** [review]
Please fix test cases and address comment above.

**Comment by AilinKid** [comment]
duplicated with https://github.com/pingcap/tidb/pull/22232, I have merged these together.

**Pull request closed without merging**